spring-integration-file/src/main/java/org/springframework/integration/file/remote/synchronizer/AbstractInboundFileSynchronizer.java

```diff
@@ -164,6 +169,7 @@ public void setTemporaryFileSuffix(String temporaryFileSuffix) {
 	 */
 	public void setRemoteDirectory(String remoteDirectory) {
 		this.remoteDirectoryExpression = new LiteralExpression(remoteDirectory);
+		evalueRemoteDirectory();
```

Comment:
Is it really safe to do this?
Or are we going to document that `RotatingServerAdvice` should not be used together with the `fixedRate` or some tight cron?

Reply:
This is actually a problem for all smart pollers - will add a warn log and docs.

Reply:
To close the topic I meant this yesterday:
````java
EvaluationContext context = this.evaluationContext;
if (!(this.localFilenameGeneratorExpression instanceof LiteralExpression) &&
					this.localFilenameGeneratorExpression.getExpressionString().contains("#remoteDirectory")) {
			context = IntegrationContextUtils.getEvaluationContext(this.beanFactory);
			context.setVariable("remoteDirectory", remoteDirectoryPath);
}

return this.localFilenameGeneratorExpression.getValue(context , remoteFileName,	String.class);
````

But now you have done fully different solution and it need its own special review.

For example how about to synchronize in the `RotatingServerAdvice` when we reconfigure a `MessageSource`, so only the last one will apply its state? We are about reconfiguring source anyway, so does it really matter which one wins?

Reply:
Well, if we have a fast poller and fair rotation, the first poller might get the directory from the second rotation. Your comment yesterday made me realize that all advices that can mutate the MS should not have a TE.

Reply:
Well, sorry for confusing: I didn't think to broad in that place...
And actually the problem not only with the `TE`. The low `fixedRate` (including similar cron) may cause similar race condition.

We definitely should document (or even just reject otherwise) a limitation of smart advices for only a single thread. Otherwise we indeed can end up with the stale state when one thread deals with the options from another.

Reply:
>The low fixedRate (including similar cron) may cause similar race condition.

That is not true; without a TE the next poll won't be scheduled until the current one completes; see 

```java
	/**
	 * Determine the next execution time according to the given trigger context.
	 * @param triggerContext context object encapsulating last execution times
	 * and last completion time
	 * @return the next execution time as defined by the trigger,
	 * or {@code null} if the trigger won't fire anymore
	 */
	@Nullable
	Date nextExecutionTime(TriggerContext triggerContext);
```

Even with `fixedRate`, the actual start of the next poll doesn't happen until this one finishes (fixedRate just means it's based on the start time, not the end time). If the run time is > fixedRate then the next one runs immediately.

Reply:
Yeah... I see what you mean - `ReschedulingRunnable`:
```
public void run() {
	Date actualExecutionTime = new Date();
	super.run();
	Date completionTime = new Date();
	synchronized (this.triggerContextMonitor) {
		Assert.state(this.scheduledExecutionTime != null, "No scheduled execution");
		this.triggerContext.update(this.scheduledExecutionTime, actualExecutionTime, completionTime);
		if (!obtainCurrentFuture().isCancelled()) {
			schedule();
		}
	}
}
```

so, we definitely don't run `run();` until the `super.run();` is done.